File: 0501-1000/0523. Continuous Subarray Sum.java
// Medium
// Map
// Time:O(n), Space:O(min(n,k))
// https://leetcode.cn/problems/continuous-subarray-sum/

import java.util.*;

class Solution {
    public boolean checkSubarraySum(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);
        int sum = 0;

        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            int mod = sum % k;

            if (map.containsKey(mod)) {
                if (i - map.get(mod) >= 2)
                    return true;
            } else {
                map.put(mod, i);
            }
        }
        return false;
    }
}

File: 0501-1000/0523. Continuous Subarray Sum.py
# Time:O(n), Space:O(min(n,k))

from typing import List

class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        mod_map = {0: -1} # mod -> first appear idx
        prefix = 0

        for i, num in enumerate(nums):
            prefix += num

            if k != 0:
                prefix %= k
            
            if prefix in mod_map:
                # len should >= 2
                if i - mod_map[prefix] > 1:
                    return True
            else:
                mod_map[prefix] = i
        
        return False
    